storage/innobase/log/log0online.cc

```diff
@@ -224,6 +222,9 @@ class log_parse_buffer final : public log_buffer<RECV_PARSING_BUF_SIZE> {
   parse_result parse_status{parse_result::OK};
   iterator end_ptr{buffer.begin()};
   lsn_t end_lsn{0};
+  lsn_t parse_start_lsn{0};
+  lsn_t parse_stop_lsn{0};
```

Comment:
can you explain briefly how we use these two LSNs to determine the exact stop_LSN (at or around checkpoint_LSN?)

Reply:
With the recent changes to redo log in 8.0 checkpoint LSN may point into a middle of log record. So now for both parse stop and start points we need to find first record group which starts at or after checkpoint LSN.
The parse_start_lsn and parse_stop_lsn are set to LSNs corresponding to those record groups starting points.

Once we read a block containing checkpoint LSN for parse start point, we check first record group offset for that block (log_block_get_first_rec_group()). In case this offset points after starting checkpoint within this block - we found parse_start_lsn. When parsing data everything with LSN less than parse_start_lsn will be skipped.

A bit different scenario is when first record group offset for the block points before starting checkpoint - we can not determine actual parse_start_lsn only from reading offset from block header, this can be done only while parsing data. In parse buffer we see that parse_start_lsn isn't set so we parse records one after another and based on parsed record length and their type we determine first record group located after starting checkpoint LSN. That is our parse_start_lsn in this case. Data from those parsed blocks located before parse_start_lsn is skipped.

The similar approach is used for parse_stop_lsn. We start searching for it once we reach a block containing ending  checkpoint LSN. We continue parsing records after checkpoint LSN until parse_stop_lsn is reached.

In addition, parse_start_lsn and parse_stop_lsn are not necessarily located in the same log block as corresponding checkpoint LSN. It may be in the next block.

Reply:
Thanks for a detailed explanation, I think it deserves to be in code (and maybe parts of it in the commit message) as well.